A number of fixes

raphlinus opened the pull request:
Set GPIO pins to the appropriate mode rather than leaving it disabled.

Make PIO setup more correct, including 20 bits per FIFO pull, unity clock divisor, and setting FIFOs to Tx direction, shift direction right. Set pindir.

Wait for DMA config to load. Put symbols in RAM.

Fix bug in TMDS encoding.

raphlinus committed:
A number of fixes

Set GPIO pins to the appropriate mode rather than leaving it disabled.

Make PIO setup more correct, including 20 bits per FIFO pull, unity clock divisor, and setting FIFOs to Tx direction, shift direction right. Set pindir.

Wait for DMA config to load. Put symbols in RAM.

Fix bug in TMDS encoding.

raphlinus commented:
This commit moves us quite a bit closer to actually getting output, but we're not quite there yet. I have verified that the dma list is being serialized out for one scan line, but what is not happening is the interrupt handler (`DMA_IRQ_0`) being called. It is in the vector table (verified by objdump), and I am calling `listen_irq0()` in `DmaChannels::start()`, so I'm not yet seeing what's wrong.

raphlinus committed:
Unmask interrupt in NVIC

raphlinus commented:
As of the latest commit (thanks thejpster for the tip!), it is calling the interrupt handler, and that seems to be mostly reloading the dma configs as expected, but it seems intermittent and not reliable. (I'm doing all my debugging by sending pio output to the LED and looking at that, with various frequency dividers)

It feels fairly close to producing blank output, but I'm not quite sure what to do next to track this down.

raphlinus commented:
I suspect that the major remaining problem is clearing the interrupt in the interrupt handler. I think what's happening in the current code is that the interrupt is not getting cleared so the handler just repeatedly gets called, causing it to cycle rapidly through vertical timing and essentially outputting a random scanline.

However, experimenting with this seems to cause the interrupt to only trigger a single time.

DusterTheFirst commented:
I am again impressed by the amount of progress you have made, especially only having an LED as a debugging tool! I plan to set up a [defmt-test](https://github.com/knurling-rs/defmt/tree/main/firmware/defmt-test) based test harness to be able to assert that the implementation behaves as it should. Before setting that up I will try to run the changes from this PR with some defmt logging and see if I can track down the problems with the interrupt handler.

I have always quite liked svd2rust's way of implementing register access, allowing lots of register operations ot be done in safe rust code, but from your comments in `dma.rs` I have started to look into chiptool as well. Do you have any strong opinion preferring the changes made by chiptool for this implementation?

raphlinus committed:
Clear interrupts

You have to clear interrupts in an interrupt handler, otherwise the interrupt controller will keep calling the handler in a tight loop.

On my board, as of this commit, the monitor syncs to the output and I see a yellow-and-red digital snow pattern, which suggests that the blue channel is being generated correctly; it also responds to the first line of EMPTY_SCANLINE_TMDS being changed.

When testing with the LED at reduced frequencies, I've observed that the wait_for_load calls on lanes 1 and 2 seem to hang, but lane 0 seems to be correct. However, this doesn't seem to be a problem at full speed.

raphlinus commented:
Thanks for the encouragement! I now have the display syncing and apparently showing the blue channel correctly, and the other two showing a fascinating form of digital noise, a bit like the color of television, tuned to a dead channel, but somehow even more cyberpunk.

I am quite fine with moving to chiptool or really any other changes to improve code cleanliness, but I do have one request: that these changes be done against a working (even if only solid color) version of the code, so we have some hope of tracking down regressions.

raphlinus committed:
Wait for FIFOs to fill before starting SMs

To try to keep the SMs in sync, we wait for the FIFOs to become full before starting them. This doesn't get to correct output though.
